README.md
# Data Engineering

## 01 - Query Project
In the Query Project, I will practice using SQL while learning about Google Cloud Platform (GCP) and BiqQuery. I'll answer business-driven questions using public datasets housed in GCP. I will use the web UI (BiqQuery) and the command-line tools, while working with them in Jupyter Notebooks.

## 02 - Tracking User Activity
I work at an ed tech firm. I've created a service that delivers assessments, and now several different customers (e.g., Pearson) want to publish their assessments on it. In this project I prepare the assessment data for data scientists who work for these customers to run queries.

## 03 - Understanding User Behavior
I am a data scientist at a game development company. The game that I developed has two events: purchase a sword and join a guild. Each event has metadata that can be quieried to gain business insights.


02_kafka-spark-hdfs/history.txt
    1  ls
    2  cd project-1-samueljgomez
    3  ls
    4  git status
    5  git remove
    6  git
    7  git reset
    8  git status
    9  git pull upstream master
   10  git status
   11  git rm .README.md.swp
   12  git status
   13  git rm .README.md.swp
   14  git status
   15  ls
   16  git add .
   17  git status
   18  git rm .README.md.swp
   19  git status
   20  git reset
   21  git status
   22  ls
   23  docker pull midsw205/base
   24  ls
   25  git status
   26  clear
   27  cd ..
   28  ls
   29  git status
   30  pwd
   31  ls
   32  cd project-1-samueljgomez
   33  ls
   34  git status
   35  git clean -fX
   36  git status
   37  ls
   38  cd ..
   39  ls
   40  pwd
   41  ls
   42  mkd w205
   43  mkdir w205
   44  cd w205
   45  man 
   46  man docker
   47  docker --help
   48  clear
   49  docker run help
   50  docker run --help
   51  clear
   52  docker run -it --rm -v ~/w205:/w205 midsw205/base bash
   53  ls
   54  cd project-1-samueljgomez
   55  cd ..
   56  cd project-1-samueljgomez/
   57  ls
   58  git status
   59  ls -lah
   60  vim README.md
   61  ls -lah
   62  git status
   63  rm .README.md.swp
   64  git status
   65  clear
   66  git branch project1
   67  git checkout project1
   68  git branch
   69  git commit README.md -m "first commit"
   70  git config --global user.name "samueljgomez"
   71  git commit README.md -m "first commit"
   72  git push
   73  git push --set-upstream origin project1
   74  git status
   75  git branch
   76  git branch -D fourthbranch
   77  git branch
   78  git
   79  git clone https://github.com/mids-w205-schioberg/course-content.git
   80  git clone https://github.com/mids-w205-schioberg/assignment-1-samueljgomez.git
   81  git
   82  sl
   83  which git
   84  pwd
   85  ls
   86  git clone https://github.com/mids-w205-schioberg/project-1-samueljgomez.git
   87  ls
   88  cd project-1-samueljgomez
   89  ls
   90  vim README.md
   91  cd project-1-samueljgomez
   92  git status
   93  ls
   94  git branch fourthbranch
   95  git status
   96  git branch
   97  git push
   98  ls -lah
   99  cd.git
  100  cd .git
  101  vim config
  102  cd ..
  103  ls'
  104  q
  105  pwd
  106  ls
  107  cd w205
  108  docker run -it --rn w205/base
  109  docker run -it --rm w205/base
  110  docker run -it --rm -v ~/w205:w205/base bash
  111  docker run -it --rm -v ~/w205: w205/base bash
  112  docker run -it --rm -v ~/w205:/w205 midsw205/base bash
  113  ls
  114  docker
  115  docker ps
  116  clear
  117  docker images
  118  docker run -it --rm -v ~/w205:/w205 midsw205/base bash
  119  docker run -it --rm -v ~/w205:/w205 midsw205/base bash which kafkacat
  120  doker  run redis
  121  docker run redis
  122  docker ps
  123  docker run -d redis
  124  docker ps
  125  docker stop proxy-agent
  126  ls
  127  cd w205
  128  ls
  129  man curl
  130  curl -L -o annot_fpid.json https://goo.gl/qWiu7d
  131  curl -L -o lp_data.csv https://goo.gl/FDFPYB
  132  ls -lah
  133  head lp_data.csv
  134  tail lp_data.csv
  135  head -n 2 lp_data.csv 
  136  cat lp_data.csv 
  137  cat lp_data.csv | sort -m
  138  cat lp_data.csv | sort -man
  139  cat lp_data.csv | sort -m
  140  ls
  141  bq query --use legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data.san_franciso.bikeshare_status`'
  142  bq query --use_legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data.san_franciso.bikeshare_status`'
  143  bq query --use_legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data:san_franciso.bikeshare_status`'
  144  bq query --use_legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data:san_francisco.bikeshare_status`'
  145  bq query --use_legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data.san_francisco.bikeshare_status`'
  146  ls
  147  bq query --use_legacy_sql=false 'SELECT count(*) FROM `bigquery-public-data.san_francisco.bikeshare_status`'
  148  tail lp_data.csv
  149  cat lp_data.csv | sort -m
  150  cat lp_data.csv | sort
  151  man
  152  yes
  153  q
  154  `
  155  fjkjdfkjlfdkjldkflkdlfjkl;dsa;
  156  ls
  157  cd w205
  158  sudo apt update
  159  sudo apt install docker-compose
  160  clear
  161  mkdir redis-standalone
  162  ls
  163  cd redis-standalone
  164  cp ../course-content/05-Storing-Data-II/example-0-docker-compose.yml docker-compose.yml
  165  vim docker-compose.yml
  166  docker run redis
  167  docker ps
  168  docker-compose up -d 
  169  docker-compose logs redis
  170  docker compose ps
  171  docker-compose ps
  172  pip install redis
  173  ipython
  174  docker-compose down
  175  docker-compose ps
  176  mkdir redis-cluster/
  177  cd redis-cluster
  178  docker-compose exec mids bash
  179  docker-compose down
  180  cd
  181  cd w205
  182  docker ps
  183  clear
  184  mkdir jupyter
  185  cd jupyter
  186  cd ..
  187  ls
  188  cd course-content/
  189  ls
  190  cd..
  191  cd ..
  192  cd jupyter
  193  cp ../course-content/05-Storing-Data-II/example-2-docker-compose.yml docker-compose.yml
  194  cd ..
  195  ls
  196  mkdir jupyternotebook
  197  ls
  198  rmdir jupyter
  199  cd jupyter
  200  ls
  201  cd ..
  202  ls
  203  cd project_01/
  204  ls
  205  cd ..
  206  cd assignment-1-samueljgomez/
  207  git branch assignment
  208  ls
  209  git checkout assignment
  210  git checkout project1
  211  cd ..
  212  cd assignment-1-samueljgomez/
  213  ls
  214  cat(README.md)
  215  pip install grip
  216  grip
  217  ls
  218  git fetch
  219  git pull remote origin
  220  git pull remote upstream
  221  ls
  222  cd
  223  ls
  224  cd w205
  225  ls
  226  cd project-1-samueljgomez/
  227  ls
  228  git pull remote upstream
  229  clear
  230  git branch
  231  git pull upstream maste
  232  git pull upstream master
  233  git pull origin master
  234  git branch
  235  ipython notebook
  236  clear
  237  ls
  238  cat README.md
  239  clear
  240  git branch -a
  241  git fetch
  242  cd
  243  ls
  244  cd w205
  245  ls

03_/write_guild_stream.py
#!/usr/bin/env python
"""Extract events from kafka and write them to hdfs
"""
# Import packages
import json
from pyspark.sql import SparkSession
from pyspark.sql.functions import udf, from_json
from pyspark.sql.types import StructType, StructField, StringType

# I am setting up the schema definition for the join guild event.
# Notice I added the guild field to track the guild the user is in.
def join_guild_event_schema():

    return StructType([
        StructField("Accept", StringType(), True),
        StructField("Host", StringType(), True),
        StructField("User-Agent", StringType(), True),
        StructField("event_type", StringType(), True),
        StructField("guild", StringType(), True)
    ])

# This is used for filtering the event type in the main() function.
@udf('boolean')
def is_guild_join(event_as_json):
    """udf for filtering events
    """
    event = json.loads(event_as_json)
    if event['event_type'] == 'join_guild':
        return True
    return False

# This is the main function of the program.
def main():
    """main
    """
    # This block of code creates a spark session.  In order to submit a job to spark we need a spark context.
    # Spark Session provides convenience and flexibility when setting up a spark context.
    # The name of the spark session is ExtractEventsJob.  The last bit of code creates the session.
    spark = SparkSession \
        .builder \
        .appName("ExtractEventsJob") \
        .getOrCreate()

    # Read from kafka using the spark readStream command.  The format is Kafka and the name of the topic is events.
    raw_events = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", "kafka:29092") \
        .option("subscribe", "events") \
        .load()

    # This block of code is doing the event filtering and casting it to strings in one step.
    # The join guild event is created as strings with components raw_event, timestamp, and any json data included in the event.
    # The schema is infered from join_guild_event_schema() (see above).
    guild_joins = raw_events \
        .filter(is_guild_join(raw_events.value.cast('string'))) \
        .select(raw_events.value.cast('string').alias('raw_event'),
                raw_events.timestamp.cast('string'),
                from_json(raw_events.value.cast('string'),
                          join_guild_event_schema()).alias('json')) \
        .select('raw_event', 'timestamp', 'json.*')

    # I am defining a sink for the data and I am writing in a stream to HDFS using a parquet format.
    # A checkpoint location is setup to prevent losing data in case of a failure.
    # The path for this table is specified in option two--"/tmp/guild_joins".
    # The sink time is set for ten seconds.
    sink = guild_joins \
        .writeStream \
        .format("parquet") \
        .option("checkpointLocation", "/tmp/checkpoints_for_guild_joins") \
        .option("path", "/tmp/guild_joins") \
        .trigger(processingTime="10 seconds") \
        .start()

    # Sink will continue to run unless stopped by sink.stop() or by exception.
    sink.awaitTermination()


if __name__ == "__main__":
    main()


03_/write_swords_stream.py
#!/usr/bin/env python
"""Extract events from kafka and write them to hdfs
"""
# Import packages
import json
from pyspark.sql import SparkSession
from pyspark.sql.functions import udf, from_json
from pyspark.sql.types import StructType, StructField, StringType, IntegerType

# I am setting up the schema definition for the purchase sword event.
# Notice I added the sword type and sword cost fields to track specific event data.
def purchase_sword_event_schema():

    return StructType([
        StructField("Accept", StringType(), True),
        StructField("Host", StringType(), True),
        StructField("User-Agent", StringType(), True),
        StructField("event_type", StringType(), True),
        StructField("sword_type", StringType(), True),
        StructField("sword_cost", IntegerType(), True)
    ])

# This is used for filtering the event type in the main() function.
@udf('boolean')
def is_sword_purchase(event_as_json):
    """udf for filtering events
    """
    event = json.loads(event_as_json)
    if event['event_type'] == 'purchase_sword':
        return True
    return False

# This is the main function of the program.
def main():
    """main
    """
    # This block of code creates a spark session.  In order to submit a job to spark we need a spark context.
    # Spark Session provides convenience and flexibility when setting up a spark context.
    # The name of the spark session is ExtractEventsJob.  The last bit of code creates the session.
    spark = SparkSession \
        .builder \
        .appName("ExtractEventsJob") \
        .getOrCreate()

    # Read from kafka using the spark readStream command.  The format is Kafka and the name of the topic is events.
    raw_events = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", "kafka:29092") \
        .option("subscribe", "events") \
        .load()

    # This block of code is doing the event filtering and casting it to strings in one step.
    # The purchase a sword event is created as strings with components raw_event, timestamp, and any json data included in the event.
    # The schema is infered from purchase_sword_event_schema() (see above).
    sword_purchases = raw_events \
        .filter(is_sword_purchase(raw_events.value.cast('string'))) \
        .select(raw_events.value.cast('string').alias('raw_event'),
                raw_events.timestamp.cast('string'),
                from_json(raw_events.value.cast('string'),
                          purchase_sword_event_schema()).alias('json')) \
        .select('raw_event', 'timestamp', 'json.*')

    # I am defining a sink for the data and I am writing in a stream to HDFS using a parquet format.
    # A checkpoint location is setup to prevent losing data in case of a failure.
    # The path for this table is specified in option two--"/tmp/sword_purchases".
    # The sink time is set for ten seconds.
    sink = sword_purchases \
        .writeStream \
        .format("parquet") \
        .option("checkpointLocation", "/tmp/checkpoints_for_sword_purchases") \
        .option("path", "/tmp/sword_purchases") \
        .trigger(processingTime="10 seconds") \
        .start()

    sink.awaitTermination()


if __name__ == "__main__":
    main()


03_web-server-event-streaming-to-database/game_api.py
# Setting up the file to execute in Python
#!/usr/bin/env python
import json
# Added to log events to Kafka.
from kafka import KafkaProducer
from flask import Flask, request
import random
# create a new flask instance
app = Flask(__name__)
producer = KafkaProducer(bootstrap_servers='kafka:29092')

# Logs events to Kafka.  The default for encode is utf-8.
def log_to_kafka(topic, event):
    event.update(request.headers)
    producer.send(topic, json.dumps(event).encode())

##################################################
# Functions
##################################################
# Function that randomly assigns sword purchases.
# The base sword is a templer sword.  There are different odds for purchasing other swords.
def sword_type():
    sword = "templer"
    rand = random.randint(1,100)
    if rand % 5 == 0:
        sword = "crusader"
        if rand % 25 == 0:
            sword = "hattori hanzo"
            if rand % 50 == 0:
                sword = "light saber"
    return sword

# This function attaches the cost of each sword to the purchase sword event.
def sword_cost(sword_type):
    if sword_type == "templer":
        cost = 5
    elif sword_type == "crusader":
        cost = 10
    elif sword_type == "hattori hanzo":
        cost = 25
    elif sword_type == "light saber":
        cost = 50
    else:
        cost = 0
    return cost

# Function for users to join a guild.
# There are five different guilds in the game.
def guild_assign():
    rand = random.randint(1,5)
    if rand == 1:
        guild = "Triad"
    elif rand == 2:
        guild = "Hellforge"
    elif rand == 3:
        guild = "Blueblades"
    elif rand == 4:
        guild = "Bouldergarde"
    else:
        guild = "Madvale"
    return guild

##################################################
# Events
##################################################
# Flask works with the app.route decorators.
# Each app.route decorator has a different name to distinguish different events.
# Inside each function we log the event to Kafka.  Inside log_to_kafka the topic name is specified.
# The business logic is returning the response.

# This is the default route. When the route is called the function is executed.
# I did not create a stream for this event, but I left it in the game for future use.
@app.route("/")
def default_response():
    default_event = {'event_type': 'default'}
    log_to_kafka('events', default_event)
    return "This is the default response!\n"

# This is the sword route, it has an event type, sword type, and sword cost properties.
# When the sword is purchased it is assigned to a specific user.
# The function outputs the response "Sword Puchased!" to the terminal.
@app.route("/purchase_a_sword")
def purchase_a_sword():
    dPSword1 = {'event_type': 'purchase_sword',
                'sword_type': sword_type()}
    dPSword2 = {'sword_cost': sword_cost(dPSword1['sword_type'])}
    purchase_sword_event = dPSword1.copy()
    purchase_sword_event.update(dPSword2)
    log_to_kafka('events', purchase_sword_event)
    return "Sword Purchased!\n"

# This is the knife route that is not used but available for future use.
@app.route("/purchase_a_knife")
def purchase_a_knife():
    purchase_knife_event = {'event_type': 'purchase_knife',
                            'description': 'very sharp knife'}
    log_to_kafka('events', purchase_knife_event)
    return "Knife Purchased!\n"

# This is the join guild route, it has an event type and a guild (name of the guild) properties. 
@app.route("/join_guild")
def join_guild():
    guild = guild_assign()
    join_guild_event = {'event_type': 'join_guild',
                        'guild':guild}
    log_to_kafka('events', join_guild_event)
    text1 = "You have joined "
    text2 = "!"
    text0 = text1 + guild + text2
    return text0 + "\n"

